### PR TITLE
[BAHIR-183]Using HDFS for saving message for mqtt source.

### PR DESCRIPTION
**Pull request by yanlin-Lynn**
Currently in spark-sql-streaming-mqtt, the received mqtt message is saved in a local file by driver, this will have the risks of losing data for cluster mode when application master failover occurs. So add a hdfs-based mqtt source to solve this problem.

### CODE REVIEW

File: sql-streaming-mqtt/src/main/scala/org/apache/bahir/sql/streaming/mqtt/HDFSMQTTSourceProvider.scala [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+  override def createSource(sqlContext: SQLContext, metadataPath: String,
+    schema: Option[StructType], providerName: String, parameters: Map[String, String]): Source = {
+
+    def e(s: String) = new IllegalArgumentException(s)
+
+    val caseInsensitiveParameter = parameters.map{case (key: String, value: String) =>
+      key.toLowerCase(Locale.ROOT) -> value
+    }
+
+    val brokerUrl: String = getParameterValue(caseInsensitiveParameter, MQTTConfig.brokerUrl, "")
```
Comment (lukasz-antoniak) [inline sql-streaming-mqtt/src/main/scala/org/apache/bahir/sql/streaming/mqtt/HDFSMQTTSourceProvider.scala]: We already have a common function to parse configuration parameters (see `MQTTUtils`), so I would try to reuse/adopt it.

File: sql-streaming-mqtt/src/main/scala/org/apache/bahir/sql/streaming/mqtt/HDFSMQTTSourceProvider.scala [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+      parameters: Map[String, String],
+      parameterName: String,
+      defaultValue: String): String = {
+
+    parameters.getOrElse(parameterName.toLowerCase(Locale.ROOT), defaultValue)
+  }
+}
+
+object HDFSMQTTSourceProvider {
+  val SCHEMA_DEFAULT = StructType(
```
Comment (lukasz-antoniak) [inline sql-streaming-mqtt/src/main/scala/org/apache/bahir/sql/streaming/mqtt/HDFSMQTTSourceProvider.scala]: There is already a schema for MQTT messages in `MQTTStreamConstants` object. Why not reuse it?

**Comment by lukasz-antoniak** [comment]
Would it be possible to implement integration tests with embedded HDFS just to make sure we do not run into regression issues?

File: sql-streaming-mqtt/src/main/scala/org/apache/bahir/sql/streaming/mqtt/HDFSMQTTSourceProvider.scala (hunk 1 above)
Comment (yanlin-Lynn) [reply sql-streaming-mqtt/src/main/scala/org/apache/bahir/sql/streaming/mqtt/HDFSMQTTSourceProvider.scala]: Thanks for your time for reviewing.  I will adjust it later!

**Comment by lresende** [comment]
@yanlin-Lynn Please rebase to latest master so I can merge.

**Comment by yanlin-Lynn** [comment]
I have done some wrong operation, close this patch and I will pull a new patch again.

**Pull request closed without merging**